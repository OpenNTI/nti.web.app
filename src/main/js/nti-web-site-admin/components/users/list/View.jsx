import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import {scoped} from 'nti-lib-locale';
import {searchable, contextual, ContextIndicator} from 'nti-web-search';
import {Loading, EmptyState} from 'nti-web-commons';
import {LinkTo} from 'nti-web-routing';// eslint-disable-line

import Store from './Store';
import Item from './Item';

const DEFAULT_TEXT = {
	loadNextPage: 'Load More',
	users: 'Users',
	empty: 'No Users',
	emptySearch: 'No Users found. Please refine your search.',
	shortSearch: 'Too many results. Please refine your search.',
	backLabel: 'View All Users',
	error: 'Unable to load Users.'
};
const t = scoped('nti-site-admin.users.list', DEFAULT_TEXT);

const store = new Store();
const propMap = {
	items: 'items',
	searchTerm: 'searchTerm',
	loading: 'loading',
	hasNextPage: 'hasNextPage',
	loadingNextPage: 'loadingNextPage',
	error: 'error'
};

@contextual(t('users'))
@searchable(store, propMap)
export default class UserListView extends React.Component {
	static propTypes = {
		items: PropTypes.array,
		searchTerm: PropTypes.string,
		loading: PropTypes.bool,
		hasNextPage: PropTypes.bool,
		loadingNextPage: PropTypes.bool,
		error: PropTypes.any
	}


	componentDidMount () {
		store.load();
	}


	onLoadNextPage = () => {
		const {hasNextPage, loadingNextPage} = this.props;

		if (hasNextPage && !loadingNextPage) {
			store.loadNextPage();
		}
	}


	render () {
		const {loading, error} = this.props;

		return (
			<div className="site-admin-user-list">
				<ContextIndicator className="context-indicator" backLabel={t('backLabel')} />
				{loading && (<div className="loading-mask"><Loading.Mask /></div>)}
				{!loading && (this.renderItems())}
				{!loading && (this.renderLoadNext())}
				{error && (this.renderError())}
			</div>
		);
	}


	renderItems () {
		const {items} = this.props;

		if (!items.length) {
			return this.renderEmptyState();
		}

		return (
			<ul>
				{items.map((user) => {
					return (
						<li key={user.NTIID}>
							<LinkTo.Object object={user}>
								<Item item={user} />
							</LinkTo.Object>
						</li>
					);
				})}
			</ul>
		);
	}


	renderLoadNext () {
		const {hasNextPage, loadingNextPage} = this.props;

		if (!hasNextPage && !loadingNextPage) { return null; }

		return (
			<div className={cx('load-next-page', {loading: loadingNextPage})} onClick={this.onLoadNextPage}>
				<span>
					{loadingNextPage && (<Loading.Spinner white />)}
					{hasNextPage && !loadingNextPage && t('loadNextPage')}
				</span>
			</div>
		);
	}


	renderEmptyState () {
		const {searchTerm} = this.props;
		const header = searchTerm ?
			(searchTerm.length < 3 ? t('shortSearch') : t('emptySearch')) :
			t('empty');

		return (
			<EmptyState header={header} />
		);
	}


	renderError () {
		return (
			<div className="error">
				{t('error')}
			</div>
		);
	}
}
