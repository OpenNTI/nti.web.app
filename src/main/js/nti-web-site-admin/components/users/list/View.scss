@import '~nti-style-common';

.site-admin-user-list {
	padding: 0 rem-calc(128px);

	.loading-mask {
		min-height: 200px;
	}

	ul {
		list-style: none;
		padding: 0;
		margin: 0;

		li {
			border-bottom: 1px solid $border-grey-light;

			a {
				text-decoration: none;
			}
		}
	}

	.load-next-page {
		text-align: center;
		cursor: pointer;
		border-radius: 5px;
		margin-top: 0.5rem;
		background-color: $primary-blue;
		color: white;
		font-size: 0.875rem;
		line-height: rem-calc(40px);
		min-height: rem-calc(40px);
	}
}
