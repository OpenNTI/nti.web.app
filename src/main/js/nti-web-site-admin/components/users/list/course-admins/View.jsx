import React, { useEffect } from 'react';

import { ErrorMessage } from '@nti/web-core';
import { DataContext } from '@nti/web-core/data';

import { CourseAdminsStore } from './Store';
import { CourseAdminsHeader, CourseAdminsHeaderPlaceholder } from './Header';
import { CourseAdminsTable, CourseAdminsTablePlaceholder } from './Table';

const Fallback = () => (
	<>
		<CourseAdminsHeaderPlaceholder />
		<CourseAdminsTablePlaceholder />
	</>
);

export function CourseAdmins({ className, searchTerm }) {
	const store = CourseAdminsStore.useStore();

	useEffect(() => store.setSearchTerm(searchTerm), [searchTerm]);

	return (
		<div className={className}>
			<DataContext
				store={store}
				fallback={<Fallback />}
				error={
					<ErrorMessage as="div" type="subhead-one" center pv="lg" />
				}
			>
				<CourseAdminsHeader />
				<CourseAdminsTable />
			</DataContext>
		</div>
	);
}
