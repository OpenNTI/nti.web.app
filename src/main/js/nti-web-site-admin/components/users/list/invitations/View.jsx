import { useEffect } from 'react';

import { scoped } from '@nti/lib-locale';
import { ErrorMessage } from '@nti/web-core';
import { DataContext } from '@nti/web-core/data';
import { WithSearch } from '@nti/web-search';

import { InvitationsStore } from './Store';
import { InvitationsHeader, InvitationsHeaderPlaceholder } from './Header';
import { InvitationsTable, InvitationsTablePlaceholder } from './Table';
import { InvitationsFooter } from './Footer';

const t = scoped('nti-web-site-admin.components.users.list.invitations.View', {
	search: 'Invitations',
});

const Fallback = () => (
	<>
		<InvitationsHeaderPlaceholder />
		<InvitationsTablePlaceholder />
	</>
);

function InvitationsView({ className, searchTerm }) {
	const store = InvitationsStore.useStore();

	useEffect(() => store.setSearchTerm(searchTerm), [searchTerm]);

	return (
		<div className={className}>
			<DataContext
				store={store}
				fallback={<Fallback />}
				error={
					<ErrorMessage as="div" type="subhead-one" center pv="lg" />
				}
			>
				<InvitationsHeader />
				<InvitationsTable />
				<InvitationsFooter />
			</DataContext>
		</div>
	);
}

export default WithSearch(InvitationsView, { label: t('search') });
