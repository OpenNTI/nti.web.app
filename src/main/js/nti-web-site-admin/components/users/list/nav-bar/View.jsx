import React from 'react';
import PropTypes from 'prop-types';
import {scoped} from '@nti/lib-locale';
import {LinkTo} from '@nti/web-routing';

import Card from '../../../common/Card';
import Tabs from '../../../common/Tabs';
import InvitePeople from '../InvitePeople';
import Store from '../table/InvitationsStore';

const DEFAULT_TEXT = {
	learners: 'Learners',
	admins: 'Admins',
	invitations: 'Invitations',
	invitePeople: 'Invite People',
	people: 'People'
};

const t = scoped('nti-site-admin.users.list.navbar.View', DEFAULT_TEXT);

export default
@Store.connect({
	loading: 'loading',
	total: 'total'
})
class UserListNavBar extends React.Component {
	static propTypes = {
		store: PropTypes.object.isRequired,
		total: PropTypes.number,
		loading: PropTypes.bool
	}

	// componentDidMount () {
	// 	this.props.store.loadPage();
	// }

	componentWillUnmount () {
		this.props.store.setUnload();
	}

	render () {
		const {loading, total, store} = this.props;

		const hasCount = !loading && (total || total === 0);

		return (
			<Card className="site-admin-user-list-nav-bar">
				<div className="header">{t('people')}</div>
				<Tabs>
					<LinkTo.Path to="./" activeClassName="active" exact>{t('learners')}</LinkTo.Path>
					<LinkTo.Path to="./admins" activeClassName="active">{t('admins')}</LinkTo.Path>
					<LinkTo.Path to="./invitations" activeClassName="active">
						{t('invitations')}
						{hasCount && <div className="invitations-count">{total || 0}</div>}
					</LinkTo.Path>
				</Tabs>
				<div className="invite" onClick={() => {
					InvitePeople.show(store);
				}}>
					{t('invitePeople')}
				</div>
			</Card>
		);
	}
}
