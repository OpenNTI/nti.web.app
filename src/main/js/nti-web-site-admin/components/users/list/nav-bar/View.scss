@import '~@nti/style-common';

.site-admin-user-list-nav-bar {
	.site-admin-tabs {
		li {
			font-size: 14px;
			line-height: 40px;
			box-shadow: none;

			a {
				padding: 0 1rem;
				position: relative;

				.invitations-count {
					position: absolute;
					top: 6px;
					right: 8px;
					color: $tertiary-grey;
					border: solid 1px $tertiary-grey;
					border-radius: 20px;
					padding: 2px 14px;
					height: 26px;
					font-size: 12px;
					line-height: 22px;
				}

				&.active {
					box-shadow: inset 4px 0 0 0 #168fe7;

					.invitations-count {
						border: 1px solid rgba(255, 255, 255, 0.38);
						color: white;
					}
				}
			}
		}
	}

	.invite {
		color: #3fb3f6;
		font-size: 14px;
		margin: 0 1rem;
		line-height: 60px;
		cursor: pointer;
	}
}
