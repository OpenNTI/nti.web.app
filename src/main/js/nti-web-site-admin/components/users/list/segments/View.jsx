import React, { useEffect } from 'react';

import { ErrorMessage } from '@nti/web-core';
import { DataContext } from '@nti/web-core/data';

import { UserSegmentsStore } from './Store';
import { UserSegmentsFooter } from './Footer';
import { UserSegmentsHeader, UserSegmentHeaderPlaceholder } from './Header';
import { UserSegmentsTable, UserSegmentsTablePlaceholder } from './Table';

const Fallback = () => (
	<>
		<UserSegmentHeaderPlaceholder />
		<UserSegmentsTablePlaceholder />
	</>
);

export function Segments({ className, searchTerm }) {
	const store = UserSegmentsStore.useStore();

	useEffect(() => store.setSearchTerm(searchTerm), [searchTerm]);

	return (
		<div className={className}>
			<DataContext
				store={store}
				fallback={<Fallback />}
				error={
					<ErrorMessage as="div" type="subhead-one" center pv="lg" />
				}
			>
				<UserSegmentsHeader />
				<UserSegmentsTable />
				<UserSegmentsFooter />
			</DataContext>
		</div>
	);
}
