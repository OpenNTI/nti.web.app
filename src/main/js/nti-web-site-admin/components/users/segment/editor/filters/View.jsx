import { useEffect, useState, useRef } from 'react';

import { useForceUpdate, ErrorMessage } from '@nti/web-core';

import { SegmentStore } from '../../Store';

import { getType } from './types/Types';
import { getCmp } from './components/Components';

const Container = styled.div`
	padding: 14px 30px 30px;
	border-top: 1px solid var(--border-grey-light);
	min-height: 200px;
`;

export function SegmentFilters() {
	const { filterSet, setFilterSet, save } = SegmentStore.useProperties();

	const seen = useRef();

	if (!seen.current) {
		seen.current = new Set();
	}

	const update = useForceUpdate();
	const [filter, setFilter] = useState(null);

	useEffect(() => {
		if (seen.current.has(filterSet)) {
			seen.current.delete(filterSet);
		} else {
			setFilter(getType(filterSet));
		}
	}, [filterSet]);

	useEffect(
		() =>
			filter?.subscribe(() => {
				if (filter.isEmpty()) {
					setFilterSet(null);
					setFilter(getType(null));
					return;
				}

				const json = filter.toJSON();
				seen.current.add(json);
				setFilterSet(json, filter.getErrors());
				update();
			}),
		[filter, setFilterSet]
	);

	const Cmp = getCmp(filter);

	if (!Cmp) {
		return null;
	}

	return (
		<Container>
			<ErrorMessage error={save.error} />
			<Cmp filter={filter} setFilter={setFilter} />
		</Container>
	);
}
