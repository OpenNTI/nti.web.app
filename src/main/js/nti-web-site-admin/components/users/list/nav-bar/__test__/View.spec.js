import renderer, { act } from 'react-test-renderer';

import * as TestUtils from '@nti/web-client/test-utils';
import { flushPromises } from '@nti/lib-commons/test-utils';

import View from '../View';

const { tearDownTestClient, setupTestClient } = TestUtils;

const getMockService = canSendInvitations => {
	return {
		getCollection: () => {
			return {
				hasLink: () => canSendInvitations,
				getLink: () => 'mockLink',
			};
		},
		getBatch: () => {
			return Promise.resolve({
				total: 5,
			});
		},
	};
};

const onBefore = canSendInvitations => {
	jest.useFakeTimers();
	setupTestClient(getMockService(canSendInvitations));
};

const onAfter = () => {
	tearDownTestClient();
};

/* eslint-env jest */
describe('Site admin user list nav bar (with invite link)', () => {
	beforeEach(() => onBefore(true));
	afterEach(onAfter);

	test('Basic render test', async () => {
		let cmp;
		act(() => {
			cmp = renderer.create(<View />);
		});

		jest.runAllTimers();
		await flushPromises();
		jest.runAllTimers();

		const tree = cmp.toJSON();

		expect(tree).toMatchSnapshot();
		cmp.unmount();
	});
});

describe('Site admin user list nav bar (without invite link)', () => {
	beforeEach(() => onBefore(false));
	afterEach(onAfter);

	test('Basic render test', async () => {
		let cmp;
		act(() => {
			cmp = renderer.create(<View />);
		});

		jest.runAllTimers();
		await flushPromises();
		jest.runAllTimers();

		const tree = cmp.toJSON();

		expect(tree).toMatchSnapshot();
		cmp.unmount();
	});
});
