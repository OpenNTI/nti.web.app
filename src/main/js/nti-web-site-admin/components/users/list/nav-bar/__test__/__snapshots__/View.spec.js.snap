// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Site admin user list nav bar (with invite link) Basic render test 1`] = `
<div
  className="site-admin-user-list-nav-bar"
>
  <div
    className="site-admin-card"
  >
    <div
      className="site-admin-tabs-container"
    >
      <ul
        className="unadorned-list site-admin-tabs"
      >
        <li>
          <div
            className="nti-link-to-path"
            exact={true}
            to="./"
          >
            Users
          </div>
        </li>
        <li>
          <div
            className="nti-link-to-path"
            to="./admins"
          >
            Site Admins
          </div>
        </li>
        <li>
          <div
            className="nti-link-to-path"
            to="./course-admins"
          >
            Course Admins
          </div>
        </li>
        <li>
          <div
            className="nti-link-to-path"
            to="./deactivated"
          >
            Deactivated
          </div>
        </li>
      </ul>
    </div>
  </div>
  <div
    className="site-admin-card"
  >
    <div
      className="invite"
      onClick={[Function]}
    >
      <i
        className="icon-addfriend"
      />
      Invite People
    </div>
    <div
      className="site-admin-tabs-container"
    >
      <ul
        className="unadorned-list site-admin-tabs"
      >
        <li>
          <div
            className="nti-link-to-path"
            to="./invitations"
          >
            Invitations
            <div
              className="invitations-count"
              data-reach-tooltip-trigger=""
              data-state="tooltip-hidden"
              onBlur={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              onMouseMove={[Function]}
              onPointerDown={[Function]}
              onPointerEnter={[Function]}
              onPointerLeave={[Function]}
              onPointerMove={[Function]}
            >
              5
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Site admin user list nav bar (without invite link) Basic render test 1`] = `
<div
  className="site-admin-user-list-nav-bar"
>
  <div
    className="site-admin-card"
  >
    <div
      className="site-admin-tabs-container"
    >
      <ul
        className="unadorned-list site-admin-tabs"
      >
        <li>
          <div
            className="nti-link-to-path"
            exact={true}
            to="./"
          >
            Users
          </div>
        </li>
        <li>
          <div
            className="nti-link-to-path"
            to="./admins"
          >
            Site Admins
          </div>
        </li>
        <li>
          <div
            className="nti-link-to-path"
            to="./course-admins"
          >
            Course Admins
          </div>
        </li>
        <li>
          <div
            className="nti-link-to-path"
            to="./deactivated"
          >
            Deactivated
          </div>
        </li>
      </ul>
    </div>
  </div>
  <div
    className="site-admin-card"
  >
    <div
      className="site-admin-tabs-container"
    >
      <ul
        className="unadorned-list site-admin-tabs"
      >
        <li>
          <div
            className="nti-link-to-path"
            to="./invitations"
          >
            Invitations
            <div
              className="invitations-count"
              data-reach-tooltip-trigger=""
              data-state="tooltip-hidden"
              onBlur={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              onMouseMove={[Function]}
              onPointerDown={[Function]}
              onPointerEnter={[Function]}
              onPointerLeave={[Function]}
              onPointerMove={[Function]}
            >
              5
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
